File: src/main/java/software/craftsmanship/debruijn/graphs/Graph.java
package software.craftsmanship.debruijn.graphs;

import java.util.HashSet;
import java.util.Set;

import software.craftsmanship.debruijn.combinatorics.Word;
import software.craftsmanship.debruijn.graphs.block.EdgeYieldBlock;
import software.craftsmanship.debruijn.graphs.block.VertexYieldBlock;

public class Graph<T> {
    private final Set<Word<T>> vertices = new HashSet<Word<T>>();
    private final Set<Edge<T>> edges = new HashSet<Edge<T>>();

    public void addVertex(Word<T> word) {
        vertices.add(word);
    }

    public void allVertices(VertexYieldBlock<T> block) {
        for (Word<T> word : vertices) {
            block.yield(word);
        }
    }

    public EdgeBuilder<T> addEdgeFrom(Word<T> source) {
        return new EdgeBuilder<T>(this, source);
    }

    private void addEdge(Edge<T> edge) {
        edges.add(edge);
    }

    public void allEdges(EdgeYieldBlock<T> block) {
        for (Edge<T> edge : edges) {
            block.yield(edge);
        }
    }

    public int numberOfEdges() {
        return edges.size();
    }

    public class EdgeBuilder<U> {
        private final Graph<U> graph;
        private final Word<U> source;

        public EdgeBuilder(Graph<U> graph, Word<U> source) {
            this.graph = graph;
            this.source = source;
        }

        public Edge<U> to(Word<U> sink) {
            Edge<U> edge = new Edge<U>(source, sink);
            graph.addEdge(edge);
            return edge;
        }

    }
}

File: lib/debruijn/graph.rb
class Graph
  def initialize
    @vertices = []
    @edges = []
  end

  def add_vertex(vertex)
    @vertices << vertex
  end

  def all_vertices
    @vertices.each { |vertex|
      yield vertex
    }
  end

  def add_edge_from(source)
    EdgeAdder.new(self, source)
  end

  def add_edge(source, sink)
    edge = Edge.new(source, sink)
    @edges << edge
    edge
  end

  def all_adges
    @edges.each { |edge|
      yield edge
    }
  end

  def number_of_edges
    @edges.length
  end
end

class Edge
  attr_reader :source, :sink
  attr_accessor :label
  def initialize(source, sink)
    @source = source
    @sink = sink
    @label = nil
  end

  def ==(edge)
    self.eql?(edge)
  end
  
  def eql?(edge)
    self.class.eql?(edge.class) && self.source.eql?(edge.source) && self.sink.eql?(edge.sink)
  end
  
  def hash()
    @source.hash ^ @sink.hash
  end
end

class EdgeAdder
  def initialize(graph, source)
    @graph = graph
    @source = source
  end

  def to(sink)
    @graph.add_edge(@source, sink)
  end
end
